librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+pub mod data;
+
+use crate::{
+    id::{uri::RadicleUri, user::User},
+    keys::device::{Key, PublicKey, Signature},
+};
+use async_trait::async_trait;
+use futures::stream::{Stream, StreamExt};
+use multihash::{Multihash, Sha2_256};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+    str::FromStr,
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Serialization failed ({0})")]
+    SerializationFailed(serde_json::error::Error),
+
+    #[error("Invalid UTF8 ({0})")]
+    InvalidUtf8(std::string::FromUtf8Error),
+
+    #[error("Invalid buffer encoding ({0})")]
+    InvalidBufferEncoding(String),
+
+    #[error("Invalid hash ({0})")]
+    InvalidHash(String),
+
+    #[error("Invalid URI ({0})")]
+    InvalidUri(String),
+
+    #[error("Signature already present ({0})")]
+    SignatureAlreadyPresent(PublicKey),
+
+    #[error("Invalid data ({0})")]
+    InvalidData(String),
+
+    #[error("Key not present ({0})")]
+    KeyNotPresent(PublicKey),
+
+    #[error("User not present ({0})")]
+    UserNotPresent(RadicleUri),
+
+    #[error("User key not present (uri {0}, key {1})")]
+    UserKeyNotPresent(RadicleUri, PublicKey),
+
+    #[error("Signature missing")]
+    SignatureMissing,
+
+    #[error("Signature decoding failed")]
+    SignatureDecodingFailed,
+
+    #[error("Signature verification failed")]
+    SignatureVerificationFailed,
+
+    #[error("Resolution failed (uri {0})")]
+    ResolutionFailed(String),
+}
+
+#[derive(Debug, Error)]
+pub enum UpdateVerificationError {
+    #[error("Non monotonic revision")]
+    NonMonotonicRevision,
+
+    #[error("Wrong parent hash")]
+    WrongParentHash,
+
+    #[error("Update without previous quorum")]
+    NoPreviousQuorum,
+
+    #[error("Update without current quorum")]
+    NoCurrentQuorum,
+}
+
+#[derive(Debug, Error)]
+pub enum HistoryVerificationError {
+    #[error("Empty history")]
+    EmptyHistory,
+
+    #[error("Error at revsion (rev {revision:?}, err {error:?})")]
+    ErrorAtRevision { revision: u64, error: Error },
+
+    #[error("Update error (rev {revision:?}, err {error:?})")]
+    UpdateError {
+        revision: u64,
+        error: UpdateVerificationError,
+    },
+}
+
+/// A type expressing *who* is signing an `Entity`:
+///
+///  * either a specific user (identified by their URN),
```

Comment:
You can move these doc comments to the enum branches themselves. I'll add a suggestion below to demonstrate.